File: 剑指offer/56.2.数组中数字出现的次数 II.py
from typing import List


class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        counts = [0] * 32
        for num in nums:
            for j in range(32):
                counts[j] += num & 1
                num >>= 1
        res, m = 0, 3
        for i in range(32):
            res <<= 1
            res |= counts[31 - i] % m
        return res if counts[31] % m == 0 else ~(res ^ 0xffffffff)

File: 剑指offer/56.2.数组中数字出现的次数 II.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int singleNumber(vector<int>& nums)
    {
        vector<int> bit(32, 0);

        for (int num : nums) {
            for (int i = 0; i < 32; i++) {
                bit[i] += num & 1;
                num >>= 1;
            }
        }

        int ans = 0;

        for (int i = 31; i >= 0; i--) {
            ans <<= 1;
            ans += bit[i] % 3;
        }

        return ans;
    }
};